File: 1996-the-number-of-weak-characters-in-the-game/1996-the-number-of-weak-characters-in-the-game.cpp
class Solution {
public:
    int numberOfWeakCharacters(vector<vector<int>>& props) {

        sort(props.begin(),props.end(),[](vector<int> &a, vector<int> &b) {
            return a[0] != b[0] ? a[0] < b[0] : a[1] > b[1];
        });
        
        vector<int> temp;
        for(int i = props.size() - 1; i >= 0; i--) {
            int ind = upper_bound(temp.begin(),temp.end(),props[i][1]) - temp.begin();
            if(ind >= temp.size()) {
                temp.push_back(props[i][1]);
            }
        }
        return props.size() - temp.size();
    }
};

File: 1996-the-number-of-weak-characters-in-the-game/1996-the-number-of-weak-characters-in-the-game.py
class Solution:
    def numberOfWeakCharacters(self, p: List[List[int]]) -> int:
        p.sort()
        pmax = p[len(p) - 1][0]
        maxi = p[len(p) - 1][1]
        n = len(p)
        lp = pmax
        ans = 0
        newMaxi = -1
        for i in range(n-2,-1,-1):
            if p[i][0] == lp:
                if lp == pmax:
                    continue
                else:
                    ans += 1 if p[i][1] < maxi else 0
                    newMaxi = max(newMaxi,p[i][1])
            else:
                maxi = max(maxi,newMaxi)
                ans += 1 if p[i][1] < maxi else 0
                lp = p[i][0]
                newMaxi = p[i][1]
                
        return ans
        